velox/functions/prestosql/window/benchmarks/PrefixSortBenchmark.cpp

```diff
@@ -0,0 +1,334 @@
+        {"k_norm", INTEGER()},
+        {"k_hash", INTEGER()},
+        {"k_sort", INTEGER()},
+        {"i32", INTEGER()},
+        {"i64", BIGINT()},
+        {"f32", REAL()},
+        {"f64", DOUBLE()},
+        {"i32_halfnull", INTEGER()},
+        {"i64_halfnull", BIGINT()},
+        {"f32_halfnull", REAL()},
+        {"f64_halfnull", DOUBLE()},
+    });
+
+    VectorFuzzer::Options opts;
+    opts.vectorSize = kRowsPerVector;
+    opts.nullRatio = 0;
+    VectorFuzzer fuzzer(opts, pool_.get(), FLAGS_fuzzer_seed);
+
+    std::vector<RowVectorPtr> vectors;
+    for (auto i = 0; i < kNumVectors; ++i) {
+      std::vector<VectorPtr> children;
+
+      // Generate key with a small number of unique values from a small range
+      // (0-16).
+      children.emplace_back(makeFlatVector<int32_t>(
+          kRowsPerVector, [](auto row) { return row % 17; }));
+
+      // Generate key with a small number of unique values from a large range
+      // (300 total values).
+      children.emplace_back(
+          makeFlatVector<int32_t>(kRowsPerVector, [](auto row) {
+            if (row % 3 == 0) {
+              return std::numeric_limits<int32_t>::max() - row % 100;
+            } else if (row % 3 == 1) {
+              return row % 100;
+            } else {
+              return std::numeric_limits<int32_t>::min() + row % 100;
+            }
+          }));
+
+      // Generate key with many unique values from a large range (500K total
+      // values).
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+
+      // Generate a column with increasing values to get a deterministic sort
+      // order.
+      children.emplace_back(makeFlatVector<int32_t>(
+          kRowsPerVector, [](auto row) { return row; }));
+
+      // Generate random values without nulls.
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+      children.emplace_back(fuzzer.fuzzFlat(BIGINT()));
+      children.emplace_back(fuzzer.fuzzFlat(REAL()));
+      children.emplace_back(fuzzer.fuzzFlat(DOUBLE()));
+
+      // Generate random values with nulls.
+
+      opts.nullRatio = 0.05; // 50%
+      fuzzer.setOptions(opts);
+
+      children.emplace_back(fuzzer.fuzzFlat(INTEGER()));
+      children.emplace_back(fuzzer.fuzzFlat(BIGINT()));
+      children.emplace_back(fuzzer.fuzzFlat(REAL()));
+      children.emplace_back(fuzzer.fuzzFlat(DOUBLE()));
+
+      vectors.emplace_back(makeRowVector(inputType_->names(), children));
+    }
+    filePath_ = TempFilePath::create();
+    writeToFile(filePath_->getPath(), vectors);
+  }
+
+  ~WindowBenchmark() override {
+    HiveConnectorTestBase::TearDown();
+  }
+
+  void TestBody() override {}
+
+  // Enhance this function to include a frame clause.
+  void run(
+      const std::string& key,
+      const std::string& aggregate,
+      bool prefixSort = true) {
+    folly::BenchmarkSuspender suspender;
+
+    std::string functionSql = fmt::format(
+        "{} over (partition by {} order by k_sort)", aggregate, key);
```

Comment:
I'd expect that multi-key sort would show more improvement. Would it be possible to try that?

Reply:
All the window sorts were inherently multi-key since they involved 1 partition key + 1 order by key. However, I added some new runs now with multiple partition keys as well.

The results are modest https://gist.github.com/aditi-pandit/4942d49d91394343f57010bc72d6c831